tests/handlers/test_federation.py

```diff
@@ -449,3 +451,117 @@ def test_invalid_nested(self) -> None:
                 },
                 RoomVersions.V6,
             )
+
+
+class PartialJoinTestCase(unittest.FederatingHomeserverTestCase):
+    def test_failed_partial_join_is_clean(self) -> None:
+        """
+        Tests that, when failing to partial-join a room, we don't get stuck with
+        a partial-state flag on a room.
+        """
+
+        fed_handler = self.hs.get_federation_handler()
+        fed_client = fed_handler.federation_client
+
+        room_id = "!room:example.com"
+        membership_event = make_event_from_dict(
+            {
+                "room_id": room_id,
+                "type": "m.room.member",
+                "sender": "@alice:test",
+                "state_key": "@alice:test",
+                "content": {"membership": "join"},
+            },
+            RoomVersions.V10,
+        )
+
+        mock_make_membership_event = Mock(
+            return_value=make_awaitable(
+                (
+                    "example.com",
+                    membership_event,
+                    RoomVersions.V10,
+                )
+            )
+        )
+
+        EVENT_CREATE = make_event_from_dict(
+            {
+                "room_id": room_id,
+                "type": "m.room.create",
+                "sender": "@kristina:example.com",
+                "state_key": "",
+                "event_id": "$e",
```

Comment:
I'm a little sad that we don't use "real" event IDs here (which `make_event_from_dict` will compute for us if we give it the right room version). Would the auth events get rejected for not having the correct event IDs?

Reply:
Hmm, interesting. I added the event ID specifically because something complained when I didn't. If you think this should work, I'll look into it

Reply:
It should be dependent on the room version. Event IDs became hashes of event content only in some later room version (V3 by the looks of things).

Reply:
aha, that makes sense. Before then they were arbitrary IIRC.

glad you mentioned it; I've updated the PR to use computed event IDs now